src/zx/FunctionalityConstruction.cpp

```diff
@@ -56,13 +58,18 @@ namespace zx {
         diag.addEdge(qubits[ctrl], qubits[target]);
     }
 
-    void FunctionalityConstruction::addCphase(ZXDiagram& diag, const PiRational& phase, Qubit ctrl, Qubit target,
-                                              std::vector<Vertex>& qubits) {
-        addZSpider(diag, ctrl, qubits, Expression(phase / 2));
+    void
+    FunctionalityConstruction::addCphase(ZXDiagram& diag, const PiExpression& phase,
+                                         Qubit ctrl, Qubit target,
+                                         std::vector<Vertex>& qubits) {
+        auto new_const = phase.getConst() / 2;
+        auto new_phase = phase / 2.0;
+        new_phase.setConst(new_const);
+        addZSpider(diag, ctrl, qubits, new_phase); //todo maybe should provide a method for int division
         addCnot(diag, ctrl, target, qubits);
-        addZSpider(diag, target, qubits, Expression(-phase / 2));
+        addZSpider(diag, target, qubits, -new_phase);
         addCnot(diag, ctrl, target, qubits);
-        addZSpider(diag, target, qubits, Expression(phase / 2));
+        addZSpider(diag, target, qubits, new_phase);
     }
 
     void FunctionalityConstruction::addSwap(ZXDiagram& diag, Qubit ctrl, Qubit target,
```

Comment:
```suggestion
    void FunctionalityConstruction::addSwap(ZXDiagram& diag, const Qubit ctrl, const Qubit target,
```

Reply:
They really thought they wouldn't get consted. But the const-police will prevail.